File: rayfin_banking_app/rayfin/functions/llm.py
from __future__ import annotations

import json
from typing import Any

from openai import AzureOpenAI

from config import settings


class AzureLLMClient:
    def __init__(self) -> None:
        self._client = (
            AzureOpenAI(
                api_key=settings.azure_openai_api_key,
                api_version=settings.azure_openai_api_version,
                azure_endpoint=settings.azure_openai_endpoint,
            )
            if settings.llm_configured
            else None
        )

    def generate_json(
        self,
        system_prompt: str,
        user_payload: dict[str, Any],
        fallback: dict[str, Any],
        temperature: float = 0.1,
    ) -> dict[str, Any]:
        if not self._client:
            return fallback

        response = self._client.chat.completions.create(
            model=settings.azure_openai_deployment,
            temperature=temperature,
            response_format={'type': 'json_object'},
            messages=[
                {'role': 'system', 'content': system_prompt},
                {'role': 'user', 'content': json.dumps(user_payload, ensure_ascii=True)},
            ],
        )
        content = response.choices[0].message.content or '{}'
        try:
            return json.loads(content)
        except json.JSONDecodeError:
            return fallback


llm_client = AzureLLMClient()

File: rayfin_banking_app/rayfin/functions/src/llm.ts
export class AzureLlmClient {
  private readonly endpoint = process.env.AZURE_OPENAI_ENDPOINT?.trim() ?? '';
  private readonly apiKey = process.env.AZURE_OPENAI_API_KEY?.trim() ?? '';
  private readonly deployment = process.env.AZURE_OPENAI_DEPLOYMENT?.trim() ?? '';
  private readonly apiVersion = process.env.AZURE_OPENAI_API_VERSION?.trim() || '2024-10-21';

  private get configured(): boolean {
    return Boolean(this.endpoint && this.apiKey && this.deployment);
  }

  async generateJson<T extends object>(
    systemPrompt: string,
    userPayload: Record<string, unknown>,
    fallback: T,
    temperature = 0.1
  ): Promise<T> {
    if (!this.configured) {
      console.warn('Azure OpenAI is not configured for the app functions; using fallback response.');
      return fallback;
    }

    const url = new URL(
      `/openai/deployments/${encodeURIComponent(this.deployment)}/chat/completions`,
      this.endpoint.endsWith('/') ? this.endpoint : `${this.endpoint}/`
    );
    url.searchParams.set('api-version', this.apiVersion);

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'api-key': this.apiKey,
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        model: this.deployment,
        temperature,
        response_format: { type: 'json_object' },
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: JSON.stringify(userPayload) },
        ],
      }),
    }).catch((error: unknown) => {
      console.warn('Azure OpenAI request failed; using fallback response.', error);
      return undefined;
    });

    if (!response) {
      return fallback;
    }

    if (!response.ok) {
      console.warn(`Azure OpenAI returned ${response.status}; using fallback response.`);
      return fallback;
    }

    const body = (await response.json().catch((error: unknown) => {
      console.warn('Azure OpenAI response was not valid JSON; using fallback response.', error);
      return undefined;
    })) as
      | {
          choices?: Array<{
            message?: {
              content?: string | null;
            };
          }>;
        }
      | undefined;

    const content = body?.choices?.[0]?.message?.content ?? '{}';

    try {
      return JSON.parse(content) as T;
    } catch (error) {
      console.warn('Azure OpenAI returned a non-JSON content payload; using fallback response.', error);
      return fallback;
    }
  }
}

export const llmClient = new AzureLlmClient();
